### PR TITLE
Config option to disable copying of outputs on cache hit for JES

### PR DESCRIPTION
**Pull request by Horneth**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -19,7 +19,7 @@ before_install:
   - openssl aes-256-cbc -K "$encrypted_5ebd3ff04788_key" -iv "$encrypted_5ebd3ff04788_iv" -in src/bin/travis/resources/jesConf.tar.enc -out jesConf.tar -d || true
 env:
   global:
-    - CENTAUR_BRANCH=develop
+    - CENTAUR_BRANCH=cromwell-2309
```
Comment (Horneth) [inline .travis.yml]: revert

File: core/src/main/resources/reference.conf [hunk 2]
```diff
@@ -507,6 +507,14 @@ backend {
     #      }
     #    }
     #
+    #    caching {
+    #      # When a cache hit is found, the following duplication strategy will be followed to use the cached outputs
+    #      # Possible values: "copy", "original". Defaults to "copy"
+    #      # "copy": Copy the output files
+    #      # "original": DO NOT copy the output files but point to the original output files instead
```
Comment (Horneth) [inline core/src/main/resources/reference.conf]: other name suggestions welcome 😄

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/callcaching/JesBackendCacheHitCopyingActor.scala [hunk 3]
```diff
@@ -1,6 +1,30 @@
...
+  override def processSimpletons(wdlValueSimpletons: Seq[WdlValueSimpleton], sourceCallRootPath: Path) = cachingStrategy match {
+    case CopyCachedOutputs => super.processSimpletons(wdlValueSimpletons, sourceCallRootPath)
+    case UseOriginalCachedOutputs =>
+      Success(WdlValueBuilder.toJobOutputs(jobDescriptor.call.task.outputs, wdlValueSimpletons) -> Set.empty[(Path, Path)])
+  }
+  
+  override def processDetritus(sourceJobDetritusFiles: Map[String, String]) = cachingStrategy match {
+    case CopyCachedOutputs => super.processDetritus(sourceJobDetritusFiles)
+    case UseOriginalCachedOutputs =>
+      TryUtil.sequenceMap(sourceJobDetritusFiles.mapValues(getPath), "Failed to make paths out of job detritus") map { _ -> Set.empty[(Path, Path)] }
```
Comment (Horneth) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/callcaching/JesBackendCacheHitCopyingActor.scala]: Not sure if we should not still copy the detritus ?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/callcaching/JesCacheHitDuplicationStrategy.scala [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package cromwell.backend.impl.jes.callcaching
+
+sealed trait JesCacheHitDuplicationStrategy
+
+case object CopyCachedOutputs extends JesCacheHitDuplicationStrategy
+case object UseOriginalCachedOutputs extends JesCacheHitDuplicationStrategy
```
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/callcaching/JesCacheHitDuplicationStrategy.scala]: \n

**Comment by mcovarr** [comment]
This doesn't appear to be checking if the referenced original files still exist.

**Comment by Horneth** [comment]
@mcovarr that's true, I think it would fail eventually anyway whenever something tries to access those files (or not if nothing does ?) ? And I thought that one of the assumption in this "no-copy" mode is that we expect the files to be relatively immutable anyway. 

But I also agree that it would be cleaner to fail the "caching" if they don't exist rather than the downstream task failing by itself.

**Comment by mcovarr** [comment]
Yeah and we can fall back to try another cache hit if one is available.

@geoffjentry feel free to leave the totem on my desk.

**Comment by geoffjentry** [comment]
@mcovarr 
54032585_seagull

File: core/src/main/resources/reference.conf (hunk 2 above)
Comment (mcovarr) [inline core/src/main/resources/reference.conf]: Perhaps `reference` or `point`?

Comment (ruchim) [inline core/src/main/resources/reference.conf]: I picture this option as `reuse`-ing the original files

**Comment by Horneth** [comment]
Due to some intense  
seagull-manager-angry-bird-w425x282 I made a doc about this

**Comment by mcovarr** [comment]
red team group photo rn

Red Team

**Comment by Horneth** [comment]
@mcovarr @ruchim second round with placeholder file and check for file existence

File: core/src/main/scala/cromwell/core/io/IoCommand.scala [hunk 5]
```diff
@@ -78,3 +78,10 @@ class IoDeleteCommand(val file: Path, val swallowIOExceptions: Boolean) extends
 class IoHashCommand(val file: Path) extends IoCommand[String] {
   override def toString = s"get hash of ${file.pathAsString}"
 }
+
+/**
+  * Get Hash value for file
+  */
+class IoTouchCommand(val file: Path) extends IoCommand[Unit] {
+  override def toString = s"touch ${file.pathAsString}"
+}
```
Comment (geoffjentry) [inline core/src/main/scala/cromwell/core/io/IoCommand.scala]: \n

File: backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala [hunk 6]
```diff
@@ -100,28 +101,50 @@ abstract class StandardCacheHitCopyingActor(val standardParams: StandardCacheHit
...
 
-  when(WaitingForCopyResponses) {
-    case Event(IoSuccess(copyCommand: IoCopyCommand, _), Some(data)) =>
-      val newData = data.remove(copyCommand)
-      if (newData.copyCommandsToWaitFor.isEmpty) succeedAndStop(data.returnCode, data.copiedJobOutputs, data.copiedDetritus)
+  when(WaitingForIoResponses) {
+    case Event(IoSuccess(command: IoCommand[_], _), Some(data)) =>
+      val newData = data.remove(command)
+      if (newData.commandsToWaitFor.isEmpty) {
+        onSuccessIoCommand(data) match {
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala]: Is it possible to not replicate the non-`None`-as-an-override pattern?

File: backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala [hunk 7]
```diff
@@ -100,28 +101,50 @@ abstract class StandardCacheHitCopyingActor(val standardParams: StandardCacheHit
...
+            sendIoCommand(successCommand)
+            goto(WaitingForOnSuccessResponse)
+          case None => succeedAndStop(data.returnCode, data.newJobOutputs, data.newDetritus)
+        }
+      }
       else stay() using Option(newData)
-    case Event(IoFailure(copyCommand: IoCopyCommand, failure), _) =>
+    case Event(IoFailure(_: IoCommand[_], failure), None) =>
+      failAndStop(failure)
+    case Event(IoFailure(_: IoCommand[_], failure), Some(data)) =>
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala]: This looks collapsible:

```
case Event(IoFailure(_: IoCommand[_], failure), Some(data)) if data.commandsToWaitFor.isNotEmpty =>
  stuffInTheElseBlock
case Event(IoFailure(_: IoCommand[_], failure), _) => failAndStop(failure)
```

File: backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala [hunk 8]
```diff
@@ -160,34 +183,42 @@ abstract class StandardCacheHitCopyingActor(val standardParams: StandardCacheHit
         throw new RuntimeException(s"${JobPaths.CallRootPathKey} wasn't found for call ${jobDescriptor.call.fullyQualifiedName}", failure)
     } get
   }
+  
+  private def ioCommandsToCopyPairs(commands: Set[IoCommand[_]]): Set[PathPair] = commands collect {
+    case copyCommad: IoCopyCommand => copyCommad.source -> copyCommad.destination
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala]: copyCommad => copyCommand

File: core/src/main/resources/reference.conf [hunk 9]
```diff
@@ -504,6 +504,14 @@ backend {
     #      gcs {
     #        # A reference to a potentially different auth for manipulating files via engine functions.
     #        auth = "application-default"
+    #
+    #        caching {
+    #          # When a cache hit is found, the following duplication strategy will be followed to use the cached outputs
+    #          # Possible values: "copy", "original". Defaults to "copy"
+    #          # "copy": Copy the output files
+    #          # "original": DO NOT copy the output files but point to the original output files instead
```
Comment (mcovarr) [inline core/src/main/resources/reference.conf]: I think either here on in the README the "touch" behavior should be documented.

File: backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala (hunk 6 above)
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala]: `onSuccessIoCommand` operates on the *old* state data?

Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala]: Is there any way this new IO command could be fused to the earlier one and the logic would branch based on their combined success or failure?  It's kind of hard to follow what's happening with this follow-on command.

Comment (Horneth) [inline backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala]: I agree that the non-`None` pattern might be a bit clunky. 
But fusing this one with the rest of the commands would be even harder to follow IMO as it would look like "send all those commands but not this last one, instead wait for all the others to come back successfully and then send that one or if it failed do that instead".
Which is kind of what this is doing but using a separate FSM state which I think makes it clearer ?

Comment (Horneth) [inline backend/src/main/scala/cromwell/backend/standard/callcaching/StandardCacheHitCopyingActor.scala]: Also now that I think about it the non-`None` pattern is how the LifecycleFactory works for any optional feature that a backend might choose to implement or not with `Option[Props]` everywhere

**Comment by Horneth** [comment]
@mcovarr ready for re-review

**Comment by mcovarr** [comment]
LGTM, apologies for all the seabirding.  👍

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/callcaching/JesBackendCacheHitCopyingActor.scala [hunk 10]
```diff
@@ -1,6 +1,53 @@
...
+  
+  private val cachingStrategy = BackendInitializationData
+    .as[JesBackendInitializationData](standardParams.backendInitializationDataOption)
+    .jesConfiguration.jesAttributes.duplicationStrategy
+  
+  override def processSimpletons(wdlValueSimpletons: Seq[WdlValueSimpleton], sourceCallRootPath: Path) = cachingStrategy match {
+    case CopyCachedOutputs => super.processSimpletons(wdlValueSimpletons, sourceCallRootPath)
+    case UseOriginalCachedOutputs =>
+      val touchCommands = wdlValueSimpletons collect {
+        case WdlValueSimpleton(_, wdlFile: WdlFile) => touchCommand(getPath(wdlFile.value).get)
```
Comment (danbills) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/callcaching/JesBackendCacheHitCopyingActor.scala]: `.get` is alarming, perhaps a comment assuring the reader that it can't fail?

File: core/src/main/resources/reference.conf [hunk 11]
```diff
@@ -504,6 +504,16 @@ backend {
     #      gcs {
     #        # A reference to a potentially different auth for manipulating files via engine functions.
     #        auth = "application-default"
+    #
+    #        caching {
+    #          # When a cache hit is found, the following duplication strategy will be followed to use the cached outputs
+    #          # Possible values: "copy", "original". Defaults to "copy"
```
Comment (ruchim) [inline core/src/main/resources/reference.conf]: "original" -> "reference"

**Comment by ruchim** [comment]
👍
